File: quartz/src/main/java/com/vitorpamplona/quartz/nip03Timestamp/ots/OpenTimestamps.java
package com.vitorpamplona.quartz.nip03Timestamp.ots;

import android.util.Log;

import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.BitcoinBlockHeaderAttestation;
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.EthereumBlockHeaderAttestation;
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.LitecoinBlockHeaderAttestation;
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.PendingAttestation;
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.TimeAttestation;
import com.vitorpamplona.quartz.nip03Timestamp.ots.exceptions.VerificationException;
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpAppend;
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpCrypto;
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpSHA256;
import com.vitorpamplona.quartz.utils.Hex;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * The main class for timestamp operations.
 */
public class OpenTimestamps {

    BitcoinExplorer explorer;
    CalendarBuilder calBuilder;


    public OpenTimestamps(BitcoinExplorer explorer, CalendarBuilder builder) {
        this.explorer = explorer;
        this.calBuilder = builder;
    }

    /**
     * Show information on a detached timestamp.
     *
     * @param detachedTimestampFile The DetachedTimestampFile ots.
     * @return the string representation of the timestamp.
     */
    public String info(DetachedTimestampFile detachedTimestampFile) {
        return info(detachedTimestampFile, false);
    }

    /**
     * Show information on a detached timestamp with verbose option.
     *
     * @param detachedTimestampFile The DetachedTimestampFile ots.
     * @param verbose               Show verbose output.
     * @return the string representation of the timestamp.
     */
    public String info(DetachedTimestampFile detachedTimestampFile, boolean verbose) {
        if (detachedTimestampFile == null) {
            return "No ots file";
        }

        String fileHash = Utils.bytesToHex(detachedTimestampFile.timestamp.msg).toLowerCase(Locale.ROOT);
        String hashOp = ((OpCrypto) detachedTimestampFile.fileHashOp)._TAG_NAME();

        String firstLine = "File " + hashOp + " hash: " + fileHash + '\n';

        return firstLine + "Timestamp:\n" + detachedTimestampFile.timestamp.strTree(0, verbose);
    }

    /**
     * Show information on a timestamp.
     *
     * @param timestamp The timestamp buffered.
     * @return the string representation of the timestamp.
     */
    public String info(Timestamp timestamp) {
        if (timestamp == null) {
            return "No timestamp";
        }

        String fileHash = Utils.bytesToHex(timestamp.msg).toLowerCase(Locale.ROOT);
        String firstLine = "Hash: " + fileHash + '\n';

        return firstLine + "Timestamp:\n" + timestamp.strTree(0);
    }

    /**
     * Create timestamp with the aid of a remote calendar. May be specified multiple times.
     *
     * @param fileTimestamp The Detached Timestamp File.
     * @return The plain array buffer of stamped.
     * @throws IOException if fileTimestamp is not valid, or the stamp procedure fails.
     */
    public Timestamp stamp(DetachedTimestampFile fileTimestamp) throws IOException {
        return stamp(fileTimestamp, null, 0);
    }

    /**
     * Create timestamp with the aid of a remote calendar. May be specified multiple times.
     *
     * @param fileTimestamp       The timestamp to stamp.
     * @param calendarsUrl        The list of calendar urls.
     * @param m                   The number of calendar to use.
     * @return The plain array buffer of stamped.
     * @throws IOException if fileTimestamp is not valid, or the stamp procedure fails.
     */
    public Timestamp stamp(DetachedTimestampFile fileTimestamp, List<String> calendarsUrl, Integer m) throws IOException {
        List<DetachedTimestampFile> fileTimestamps = new ArrayList<DetachedTimestampFile>();
        fileTimestamps.add(fileTimestamp);

        return stamp(fileTimestamps, calendarsUrl, m);
    }

    /**
     * Create timestamp with the aid of a remote calendar. May be specified multiple times.
     *
     * @param fileTimestamps      The list of timestamp to stamp.
     * @param calendarsUrl        The list of calendar urls.
     * @param m                   The number of calendar to use.
     * @return The plain array buffer of stamped.
     * @throws IOException if fileTimestamp is not valid, or the stamp procedure fails.
     */
    public Timestamp stamp(List<DetachedTimestampFile> fileTimestamps, List<String> calendarsUrl, Integer m) throws IOException {
        // Parse parameters
        if (fileTimestamps == null || fileTimestamps.size() == 0) {
            throw new IOException();
        }

        if (calendarsUrl == null || calendarsUrl.size() == 0) {
            calendarsUrl = new ArrayList<String>();
            calendarsUrl.add("https://alice.btc.calendar.opentimestamps.org");
            calendarsUrl.add("https://bob.btc.calendar.opentimestamps.org");
            calendarsUrl.add("https://finney.calendar.eternitywall.com");
        }

        if (m == null || m <= 0) {
            if (calendarsUrl.size() == 0) {
                m = 2;
            } else if (calendarsUrl.size() == 1) {
                m = 1;
            } else {
                m = calendarsUrl.size();
            }
        }

        if (m < 0 || m > calendarsUrl.size()) {
            Log.e("OpenTimestamp", "m cannot be greater than available calendar neither less or equal 0");
            throw new IOException();
        }

        // Build markle tree
        Timestamp merkleTip = makeMerkleTree(fileTimestamps);

        if (merkleTip == null) {
            throw new IOException();
        }

        // Stamping
        Timestamp resultTimestamp = create(merkleTip, calendarsUrl, m);

        if (resultTimestamp == null) {
            throw new IOException();
        }

        // Result of timestamp serialization
        if (fileTimestamps.size() == 1) {
            return fileTimestamps.get(0).timestamp;
        } else {
            return merkleTip;
        }
    }

    /**
     * Create a timestamp
     *
     * @param timestamp    The timestamp.
     * @param calendarUrls List of calendar's to use.
     * @param m            Number of calendars to use.
     * @return The created timestamp.
     */
    private Timestamp create(Timestamp timestamp, List<String> calendarUrls, Integer m) {
        int capacity = calendarUrls.size();
        ExecutorService executor = Executors.newFixedThreadPool(4);
        ArrayBlockingQueue<Optional<Timestamp>> queue = new ArrayBlockingQueue<>(capacity);

        // Submit to all public calendars
        for (final String calendarUrl : calendarUrls) {
            Log.i("OpenTimestamps", "Submitting to remote calendar " + calendarUrl);

            try {
                CalendarAsyncSubmit task = new CalendarAsyncSubmit(calendarUrl, timestamp.msg);
                task.setQueue(queue);
                executor.submit(task);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        int count = 0;

        for (count = 0; count < capacity && count < m; count++) {
            try {
                Optional<Timestamp> optionalStamp = queue.take();

                if (optionalStamp.isPresent()) {
                    try {
                        Timestamp time = optionalStamp.get();
                        timestamp.merge(time);
                        Log.i("Open", ""+ timestamp.attestations.size());
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        if (count < m) {
            Log.e("OpenTimestamp", "Failed to create timestamp: requested " + String.valueOf(m) + " attestation" + ((m > 1) ? "s" : "") + " but received only " + String.valueOf(count));
        }

        //shut down the executor service now
        executor.shutdown();

        return timestamp;
    }

    /**
     * Make Merkle Tree of detached timestamps.
     *
     * @param fileTimestamps The list of DetachedTimestampFile.
     * @return merkle tip timestamp.
     */
    public Timestamp makeMerkleTree(List<DetachedTimestampFile> fileTimestamps) {
        List<Timestamp> merkleRoots = new ArrayList<>();

        for (DetachedTimestampFile fileTimestamp : fileTimestamps) {
            byte[] bytesRandom16 = new byte[16];

            try {
                bytesRandom16 = Utils.randBytes(16);
            } catch (NoSuchAlgorithmException e) {
                e.printStackTrace();
            }

            // nonce_appended_stamp = file_timestamp.timestamp.ops.add(com.vitorpamplona.quartz.ots.op.OpAppend(os.urandom(16)))
            Timestamp nonceAppendedStamp = fileTimestamp.timestamp.add(new OpAppend(bytesRandom16));
            // merkle_root = nonce_appended_stamp.ops.add(com.vitorpamplona.quartz.ots.op.OpSHA256())
            Timestamp merkleRoot = nonceAppendedStamp.add(new OpSHA256());
            merkleRoots.add(merkleRoot);
        }

        Timestamp merkleTip = Merkle.makeMerkleTree(merkleRoots);
        return merkleTip;
    }

    /**
     * Compare and verify a detached timestamp.
     *
     * @param ots     The DetachedTimestampFile containing the proof to verify.
     * @param diggest The hash of the stamped file, in bytes
     * @return Hashmap of block heights and timestamps indexed by chain: timestamp in seconds from 1 January 1970.
     * @throws Exception if the verification procedure fails.
     */

    public HashMap<VerifyResult.Chains, VerifyResult> verify(DetachedTimestampFile ots, byte[] diggest) throws Exception {
        if (!Arrays.equals(ots.fileDigest(), diggest)) {
            Log.e("OpenTimestamp", "Expected digest " + Hex.encode(ots.fileDigest()).toLowerCase(Locale.ROOT));
            Log.e("OpenTimestamp", "File does not match original!");
            throw new Exception("File does not match original!");
        }

        return verify(ots.timestamp);
    }

    /**
     * Verify a timestamp.
     *
     * @param timestamp The timestamp.
     * @return HashMap of block heights and timestamps indexed by chain: timestamp in seconds from 1 January 1970.
     * @throws Exception if the verification procedure fails.
     */
    public HashMap<VerifyResult.Chains, VerifyResult> verify(Timestamp timestamp) throws Exception {
        HashMap<VerifyResult.Chains, VerifyResult> verifyResults = new HashMap<>();

        for (Map.Entry<byte[], TimeAttestation> item : timestamp.allAttestations().entrySet()) {
            byte[] msg = item.getKey();
            TimeAttestation attestation = item.getValue();
            VerifyResult verifyResult = null;
            VerifyResult.Chains chain = null;

            try {
                if (attestation instanceof BitcoinBlockHeaderAttestation) {
                    chain = VerifyResult.Chains.BITCOIN;
                    Long time = verify((BitcoinBlockHeaderAttestation) attestation, msg);
                    int height = ((BitcoinBlockHeaderAttestation) attestation).getHeight();
                    verifyResult = new VerifyResult(time, height);
                } else if (attestation instanceof LitecoinBlockHeaderAttestation) {
                    chain = VerifyResult.Chains.LITECOIN;
                    Long time = verify((LitecoinBlockHeaderAttestation) attestation, msg);
                    int height = ((LitecoinBlockHeaderAttestation) attestation).getHeight();
                    verifyResult = new VerifyResult(time, height);
                }

                if (verifyResult != null && verifyResults.containsKey(chain)) {
                    if (verifyResult.height < verifyResults.get(chain).height) {
                        verifyResults.put(chain, verifyResult);
                    }
                }

                if (verifyResult != null && !verifyResults.containsKey(chain)) {
                    verifyResults.put(chain, verifyResult);
                }
            } catch (VerificationException e) {
                throw e;
            } catch (Exception e) {
                String text = "";

                if (chain == VerifyResult.Chains.BITCOIN) {
                    text = BitcoinBlockHeaderAttestation.chain;
                } else if (chain == VerifyResult.Chains.LITECOIN) {
                    text = LitecoinBlockHeaderAttestation.chain;
                } else if (chain == VerifyResult.Chains.ETHEREUM) {
                    text = EthereumBlockHeaderAttestation.chain;
                } else {
                    throw e;
                }

                Log.e("OpenTimestamp", Utils.toUpperFirstLetter(text) + " verification failed: " + e.getMessage());
                throw e;
            }
        }
        return verifyResults;
    }

    /**
     * Verify an Bitcoin Block Header Attestation.
     * if the node is not reachable or it fails, uses Lite-client verification.
     *
     * @param attestation The BitcoinBlockHeaderAttestation attestation.
     * @param msg         The digest to verify.
     * @return The unix timestamp in seconds from 1 January 1970.
     * @throws VerificationException if it doesn't check the merkle root of the block.
     * @throws Exception             if the verification procedure fails.
     */
    public Long verify(BitcoinBlockHeaderAttestation attestation, byte[] msg) throws VerificationException, Exception {
        Integer height = attestation.getHeight();
        BlockHeader blockInfo;

        try {
            String blockHash = explorer.blockHash(height);
            blockInfo = explorer.block(blockHash);
            Log.i("OpenTimestamps", "Lite-client verification, assuming block " + blockHash + " is valid");
        } catch (Exception e2) {
            e2.printStackTrace();
            throw e2;
        }

        return attestation.verifyAgainstBlockheader(Utils.arrayReverse(msg), blockInfo);
    }

    /**
     * Verify an Litecoin Block Header Attestation. Litecoin verification uses only lite-client verification.
     *
     * @param attestation The LitecoinBlockHeaderAttestation attestation.
     * @param msg         The digest to verify.
     * @return The unix timestamp in seconds from 1 January 1970.
     * @throws VerificationException if it doesn't check the merkle root of the block.
     * @throws Exception             if the verification procedure fails.
     */
    public Long verify(LitecoinBlockHeaderAttestation attestation, byte[] msg) throws VerificationException, Exception {
        Integer height = attestation.getHeight();
        BlockHeader blockInfo;

        try {
            String blockHash = explorer.blockHash(height);
            Log.i("OpenTimestamps", "Lite-client verification, assuming block " + blockHash + " is valid");
            blockInfo = explorer.block(blockHash);
        } catch (Exception e2) {
            e2.printStackTrace();
            throw e2;
        }

        return attestation.verifyAgainstBlockheader(Utils.arrayReverse(msg), blockInfo);
    }

    /**
     * Upgrade a timestamp.
     *
     * @param detachedTimestamp The DetachedTimestampFile containing the proof to verify.
     * @return a boolean representing if the timestamp has changed.
     * @throws Exception if the upgrading procedure fails.
     */
    public boolean upgrade(DetachedTimestampFile detachedTimestamp) throws Exception {
        // Upgrade timestamp
        boolean changed = upgrade(detachedTimestamp.timestamp);
        return changed;
    }

    /**
     * Attempt to upgrade an incomplete timestamp to make it verifiable.
     * Note that this means if the timestamp that is already complete, False will be returned as nothing has changed.
     *
     * @param timestamp The timestamp to upgrade.
     * @return a boolean representing if the timestamp has changed.
     * @throws Exception if the upgrading procedure fails.
     */
    public boolean upgrade(Timestamp timestamp) throws Exception {
        // Check remote calendars for upgrades.
        // This time we only check PendingAttestations - we can't be as agressive.

        boolean upgraded = false;
        Set<TimeAttestation> existingAttestations = timestamp.getAttestations();

        for (Timestamp subStamp : timestamp.directlyVerified()) {
            for (TimeAttestation attestation : subStamp.attestations) {
                if (attestation instanceof PendingAttestation && !subStamp.isTimestampComplete()) {
                    String calendarUrl = new String(((PendingAttestation) attestation).getUri(), StandardCharsets.UTF_8);
                    // var calendarUrl = calendarUrls[0];
                    byte[] commitment = subStamp.msg;

                    try {
                        Calendar calendar = new Calendar(calendarUrl);
                        Timestamp upgradedStamp = upgrade(subStamp, calendar, commitment, existingAttestations);

                        try {
                            subStamp.merge(upgradedStamp);
                        } catch (Exception e) {
                            e.printStackTrace();
                        }

                        upgraded = true;
                    } catch (Exception e) {
                        Log.i("OpenTimestamps", e.getMessage());
                    }
                }
            }
        }

        return upgraded;
    }

    private Timestamp upgrade(Timestamp subStamp, Calendar calendar, byte[] commitment, Set<TimeAttestation> existingAttestations) throws Exception {
        Timestamp upgradedStamp;

        try {
            upgradedStamp = calendar.getTimestamp(commitment);

            if (upgradedStamp == null) {
                throw new Exception("Invalid stamp");
            }
        } catch (Exception e) {
            Log.i("OpenTimestamps", "Calendar " + calendar.getUrl() + ": " + e.getMessage());
            throw e;
        }

        Set<TimeAttestation> attsFromRemote = upgradedStamp.getAttestations();

        if (attsFromRemote.size() > 0) {
            Log.i("OpenTimestamps", "Got 1 attestation(s) from " + calendar.getUrl());
        }

        // Set difference from remote attestations & existing attestations
        Set<TimeAttestation> newAttestations = attsFromRemote;
        newAttestations.removeAll(existingAttestations);

        // changed & found_new_attestations
        // foundNewAttestations = true;
        // Log.i("OpenTimestamps", attsFromRemote.size + ' attestation(s) from ' + calendar.url);

        // Set union of existingAttestations & newAttestations
        existingAttestations.addAll(newAttestations);

        return upgradedStamp;
        // subStamp.merge(upgradedStamp);
        // args.cache.merge(upgraded_stamp)
        // sub_stamp.merge(upgraded_stamp)
    }
}

File: quartz/src/androidMain/kotlin/com/vitorpamplona/quartz/nip03Timestamp/ots/OpenTimestamps.kt
/**
 * Copyright (c) 2025 Vitor Pamplona
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package com.vitorpamplona.quartz.nip03Timestamp.ots

import android.util.Log
import com.vitorpamplona.quartz.nip01Core.core.toHexKey
import com.vitorpamplona.quartz.nip03Timestamp.ots.Utils.arrayReverse
import com.vitorpamplona.quartz.nip03Timestamp.ots.Utils.randBytes
import com.vitorpamplona.quartz.nip03Timestamp.ots.VerifyResult.Chains
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.BitcoinBlockHeaderAttestation
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.PendingAttestation
import com.vitorpamplona.quartz.nip03Timestamp.ots.attestation.TimeAttestation
import com.vitorpamplona.quartz.nip03Timestamp.ots.exceptions.VerificationException
import com.vitorpamplona.quartz.nip03Timestamp.ots.http.Calendar
import com.vitorpamplona.quartz.nip03Timestamp.ots.http.CalendarAsyncSubmit
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpAppend
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpCrypto
import com.vitorpamplona.quartz.nip03Timestamp.ots.op.OpSHA256
import com.vitorpamplona.quartz.utils.Hex
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.security.NoSuchAlgorithmException
import java.util.Optional
import java.util.concurrent.ArrayBlockingQueue
import java.util.concurrent.Executors

/**
 * The main class for timestamp operations.
 */
class OpenTimestamps(
    var explorer: BitcoinExplorer,
    var calBuilder: CalendarBuilder?,
) {
    /**
     * Show information on a detached timestamp with verbose option.
     *
     * @param detachedTimestampFile The DetachedTimestampFile ots.
     * @param verbose               Show verbose output.
     * @return the string representation of the timestamp.
     */
    fun info(
        detachedTimestampFile: DetachedTimestampFile?,
        verbose: Boolean = false,
    ): String {
        if (detachedTimestampFile == null) {
            return "No ots file"
        }

        val fileHash =
            detachedTimestampFile.timestamp.digest
                .toHexKey()
                .lowercase()
        val hashOp = (detachedTimestampFile.fileHashOp as OpCrypto).tagName()

        val firstLine = "File $hashOp hash: $fileHash\n"

        return firstLine + "Timestamp:\n" + detachedTimestampFile.timestamp.strTree(0, verbose)
    }

    /**
     * Show information on a timestamp.
     *
     * @param timestamp The timestamp buffered.
     * @return the string representation of the timestamp.
     */
    fun info(timestamp: Timestamp?): String {
        if (timestamp == null) {
            return "No timestamp"
        }

        val fileHash = timestamp.digest.toHexKey().lowercase()
        val firstLine = "Hash: $fileHash\n"

        return firstLine + "Timestamp:\n" + timestamp.strTree(0)
    }

    /**
     * Create timestamp with the aid of a remote calendar. May be specified multiple times.
     *
     * @param fileTimestamp       The timestamp to stamp.
     * @param calendarsUrl        The list of calendar urls.
     * @param m                   The number of calendar to use.
     * @return The plain array buffer of stamped.
     * @throws IOException if fileTimestamp is not valid, or the stamp procedure fails.
     */
    @Throws(IOException::class)
    fun stamp(
        fileTimestamp: DetachedTimestampFile?,
        calendarsUrl: MutableList<String> = mutableListOf(),
        m: Int? = 0,
    ): Timestamp {
        val fileTimestamps: MutableList<DetachedTimestampFile> = ArrayList<DetachedTimestampFile>()
        fileTimestamps.add(fileTimestamp!!)

        return stamp(fileTimestamps, calendarsUrl, m)
    }

    /**
     * Create timestamp with the aid of a remote calendar. May be specified multiple times.
     *
     * @param fileTimestamps      The list of timestamp to stamp.
     * @param calendarsUrl        The list of calendar urls.
     * @param m                   The number of calendar to use.
     * @return The plain array buffer of stamped.
     * @throws IOException if fileTimestamp is not valid, or the stamp procedure fails.
     */
    @Throws(IOException::class)
    fun stamp(
        fileTimestamps: MutableList<DetachedTimestampFile>,
        calendarsUrl: MutableList<String>,
        m: Int?,
    ): Timestamp {
        // Parse parameters
        var calendarsUrl = calendarsUrl
        var m = m
        if (fileTimestamps.isEmpty()) {
            throw IOException()
        }

        if (calendarsUrl.isEmpty()) {
            calendarsUrl = ArrayList()
            calendarsUrl.add("https://alice.btc.calendar.opentimestamps.org")
            calendarsUrl.add("https://bob.btc.calendar.opentimestamps.org")
            calendarsUrl.add("https://finney.calendar.eternitywall.com")
        }

        if (m == null || m <= 0) {
            if (calendarsUrl.size == 0) {
                m = 2
            } else if (calendarsUrl.size == 1) {
                m = 1
            } else {
                m = calendarsUrl.size
            }
        }

        if (m < 0 || m > calendarsUrl.size) {
            Log.e(
                "OpenTimestamp",
                "m cannot be greater than available calendar neither less or equal 0",
            )
            throw IOException()
        }

        // Build markle tree
        val merkleTip = makeMerkleTree(fileTimestamps)

        // Stamping
        val resultTimestamp = create(merkleTip, calendarsUrl, m)

        // Result of timestamp serialization
        if (fileTimestamps.size == 1) {
            return fileTimestamps.get(0).timestamp
        } else {
            return merkleTip
        }
    }

    /**
     * Create a timestamp
     *
     * @param timestamp    The timestamp.
     * @param calendarUrls List of calendar's to use.
     * @param m            Number of calendars to use.
     * @return The created timestamp.
     */
    private fun create(
        timestamp: Timestamp,
        calendarUrls: MutableList<String>,
        m: Int,
    ): Timestamp {
        val capacity = calendarUrls.size
        val executor = Executors.newFixedThreadPool(4)
        val queue = ArrayBlockingQueue<Optional<Timestamp>>(capacity)

        // Submit to all public calendars
        for (calendarUrl in calendarUrls) {
            Log.i("OpenTimestamps", "Submitting to remote calendar $calendarUrl")

            try {
                val task = CalendarAsyncSubmit(calendarUrl, timestamp.digest)
                task.setQueue(queue)
                executor.submit<Optional<Timestamp>>(task)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        var count = 0

        count = 0
        while (count < capacity && count < m) {
            try {
                val optionalStamp = queue.take()

                if (optionalStamp.isPresent()) {
                    try {
                        val time = optionalStamp.get()
                        timestamp.merge(time)
                        Log.i("Open", "" + timestamp.attestations.size)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
            count++
        }

        if (count < m) {
            Log.e(
                "OpenTimestamp",
                "Failed to create timestamp: requested " + m.toString() + " attestation" + (if (m > 1) "s" else "") + " but received only " + count.toString(),
            )
        }

        // shut down the executor service now
        executor.shutdown()

        return timestamp
    }

    /**
     * Make Merkle Tree of detached timestamps.
     *
     * @param fileTimestamps The list of DetachedTimestampFile.
     * @return merkle tip timestamp.
     */
    fun makeMerkleTree(fileTimestamps: MutableList<DetachedTimestampFile>): Timestamp {
        val merkleRoots: MutableList<Timestamp> = ArrayList()

        for (fileTimestamp in fileTimestamps) {
            var bytesRandom16 = ByteArray(16)

            try {
                bytesRandom16 = randBytes(16)
            } catch (e: NoSuchAlgorithmException) {
                e.printStackTrace()
            }

            val nonceAppendedStamp = fileTimestamp.timestamp.add(OpAppend(bytesRandom16))
            val merkleRoot = nonceAppendedStamp.add(OpSHA256())
            merkleRoots.add(merkleRoot)
        }

        val merkleTip: Timestamp = Merkle.makeMerkleTree(merkleRoots)!!
        return merkleTip
    }

    /**
     * Compare and verify a detached timestamp.
     *
     * @param ots     The DetachedTimestampFile containing the proof to verify.
     * @param diggest The hash of the stamped file, in bytes
     * @return Hashmap of block heights and timestamps indexed by chain: timestamp in seconds from 1 January 1970.
     * @throws Exception if the verification procedure fails.
     */
    @Throws(Exception::class)
    fun verify(
        ots: DetachedTimestampFile,
        diggest: ByteArray,
    ): HashMap<Chains, VerifyResult> {
        if (!ots.fileDigest().contentEquals(diggest)) {
            Log.e("OpenTimestamp", "Expected digest " + Hex.encode(ots.fileDigest()!!).lowercase())
            Log.e("OpenTimestamp", "File does not match original!")
            throw Exception("File does not match original!")
        }

        return verify(ots.timestamp)
    }

    /**
     * Verify a timestamp.
     *
     * @param timestamp The timestamp.
     * @return HashMap of block heights and timestamps indexed by chain: timestamp in seconds from 1 January 1970.
     * @throws Exception if the verification procedure fails.
     */
    @Throws(Exception::class)
    fun verify(timestamp: Timestamp): HashMap<Chains, VerifyResult> {
        val verifyResults = HashMap<Chains, VerifyResult>()

        for (item in timestamp.allAttestations().entries) {
            val msg = item.key
            val attestation: TimeAttestation? = item.value
            var verifyResult: VerifyResult? = null

            try {
                if (attestation is BitcoinBlockHeaderAttestation) {
                    val time = verify(attestation, msg)
                    val height = attestation.height
                    verifyResult = VerifyResult(time, height)

                    val currentResults = verifyResults.get(Chains.BITCOIN)

                    if (currentResults != null) {
                        if (verifyResult.height < currentResults.height) {
                            verifyResults.put(Chains.BITCOIN, verifyResult)
                        }
                    } else {
                        verifyResults.put(Chains.BITCOIN, verifyResult)
                    }
                }
            } catch (e: VerificationException) {
                throw e
            } catch (e: Exception) {
                Log.e("OpenTimestamp", "Verification failed: " + e.message)
                throw e
            }
        }
        return verifyResults
    }

    /**
     * Verify an Bitcoin Block Header Attestation.
     * if the node is not reachable or it fails, uses Lite-client verification.
     *
     * @param attestation The BitcoinBlockHeaderAttestation attestation.
     * @param msg         The digest to verify.
     * @return The unix timestamp in seconds from 1 January 1970.
     * @throws VerificationException if it doesn't check the merkle root of the block.
     * @throws Exception             if the verification procedure fails.
     */
    @Throws(VerificationException::class, Exception::class)
    fun verify(
        attestation: BitcoinBlockHeaderAttestation,
        msg: ByteArray,
    ): Long? {
        val height = attestation.height
        val blockInfo: BlockHeader?

        try {
            val blockHash = explorer.blockHash(height)
            blockInfo = explorer.block(blockHash)
            Log.i(
                "OpenTimestamps",
                "Lite-client verification, assuming block $blockHash is valid",
            )
        } catch (e2: Exception) {
            e2.printStackTrace()
            throw e2
        }

        return attestation.verifyAgainstBlockheader(arrayReverse(msg), blockInfo)
    }

    /**
     * Upgrade a timestamp.
     *
     * @param detachedTimestamp The DetachedTimestampFile containing the proof to verify.
     * @return a boolean representing if the timestamp has changed.
     * @throws Exception if the upgrading procedure fails.
     */
    @Throws(Exception::class)
    fun upgrade(detachedTimestamp: DetachedTimestampFile): Boolean {
        // Upgrade timestamp
        val changed = upgrade(detachedTimestamp.timestamp)
        return changed
    }

    /**
     * Attempt to upgrade an incomplete timestamp to make it verifiable.
     * Note that this means if the timestamp that is already complete, False will be returned as nothing has changed.
     *
     * @param timestamp The timestamp to upgrade.
     * @return a boolean representing if the timestamp has changed.
     * @throws Exception if the upgrading procedure fails.
     */
    @Throws(Exception::class)
    fun upgrade(timestamp: Timestamp): Boolean {
        // Check remote calendars for upgrades.
        // This time we only check PendingAttestations - we can't be as agressive.

        var upgraded = false
        val existingAttestations: MutableSet<TimeAttestation> = timestamp.getAttestations()

        for (subStamp in timestamp.directlyVerified()) {
            for (attestation in subStamp.attestations) {
                if (attestation is PendingAttestation && !subStamp.isTimestampComplete) {
                    val calendarUrl = String(attestation.uri, StandardCharsets.UTF_8)
                    // var calendarUrl = calendarUrls[0];
                    val commitment = subStamp.digest

                    try {
                        val calendar = Calendar(calendarUrl)
                        val upgradedStamp = upgrade(subStamp, calendar, commitment, existingAttestations)

                        try {
                            subStamp.merge(upgradedStamp)
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }

                        upgraded = true
                    } catch (e: Exception) {
                        Log.i("OpenTimestamps", e.message ?: "${e.javaClass.simpleName}")
                    }
                }
            }
        }

        return upgraded
    }

    @Throws(Exception::class)
    private fun upgrade(
        subStamp: Timestamp,
        calendar: Calendar,
        commitment: ByteArray,
        existingAttestations: MutableSet<TimeAttestation>,
    ): Timestamp {
        val upgradedStamp: Timestamp =
            try {
                calendar.getTimestamp(commitment)
            } catch (e: Exception) {
                Log.i("OpenTimestamps", "Calendar " + calendar.url + ": " + e.message)
                throw e
            }

        val attsFromRemote: MutableSet<TimeAttestation> = upgradedStamp.getAttestations()

        if (attsFromRemote.isNotEmpty()) {
            Log.i("OpenTimestamps", "Got 1 attestation(s) from " + calendar.url)
        }

        // Set difference from remote attestations & existing attestations
        val newAttestations = attsFromRemote.minus(existingAttestations)

        // changed & found_new_attestations
        // foundNewAttestations = true;
        // Log.i("OpenTimestamps", attsFromRemote.size + ' attestation(s) from ' + calendar.url);

        // Set union of existingAttestations & newAttestations
        existingAttestations.addAll(newAttestations)

        return upgradedStamp
    }
}
